### DISCUSSION TITLE
Dogecoin Core 1.14.6

### DISCUSSION THREAD

**Discussion by patricklodder** [dogecoin/dogecoin#3054] [General]
This is a new minor version release, including important security updates and changes to network efficiency.

**All Dogecoin Core users - miners, services, relay operators and wallet users - are strongly recommended to upgrade.**

A full description of all changes is available in the [release notes](https://github.com/dogecoin/dogecoin/blob/master/doc/release-notes.md)

### Security-related fixes

This release contains multiple security related fixes:

- The alert system has been removed and processing of alert messages has been disabled
- The transaction download system has been made more reliable
- The protocol implementation has been amended to reject buggy or malformed messages
- Memory management in events of high network traffic or when connected to extremely slow peers has been improved

### Fee Recommendation

This release changes the **recommended dust limit** for all participants on the Dogecoin network from 1 DOGE to **0.01 DOGE**. The full recommendation can be found [in the documentation](https://github.com/dogecoin/dogecoin/blob/1.14.6-dev/doc/fee-recommendation.md).

### Breaking changes

- The behavior of the `dumpwallet` and `backupwallet` RPC methods, to exclusively write to the directory specified by `-backupdir` and no longer overwrite files
- The maximum number of `addnode` records that a node will track has been limited to 800

### New features

- A new RPC method, `liststucktransactions` has been added to help wallet operators identify if any transactions are stuck (have not been mined and are not in the mempool) and enable resolution. This helps in cases where node-local fee or dust policies change, to prevent the node from sending out old transactions that got lost.
- Pruning configuration has been added to the UI, to make it easier to save diskspace for users of graphical (wallet) installations. 
- A new RPC method, `setmaxconnections`, has been introduced that enables wallet and node operators to change the maximum number of connections their node allows without needing to restart allowing for more fine-grained control of the node's network capacities.
- The graph in the UI's traffic monitor has been significantly improved by keeping track of the traffic we see in the background and then rendering it on demand, at any desired scale.
- A new RPC method, `rescan`, has been added to allow node operators to rescan the chain for wallet transactions from a certain height. This can help speed up the rescan process for keys that are known to be generated after a certain date or block height.
- A new endpoint has been added to the REST server to find the current best known block at a certain height, without having to do RPC calls. This is useful for service implementations that need to track or query blocks.

### Translation updates

- Bulgarian
- Dutch
- French
- Italian
- Korean
- Persian
- Portuguese (pt_PT)
- Tamil

### RPC API Changes

- The `verifychain` RPC command no longer accepts negative values
- `addnode` will now return error -32 when the address provided is too long
- `getpeerinfo` provides 2 new fields per peer, `addr_processed` and `addr_rate_limited`

### Dependency updates

- Updated `zlib` to 1.2.12.
- Updated `libevent` to 2.1.12-stable.
- Update the LevelDB subtree to use the same version as Bitcoin 22.0, from the [bitcoin-core maintained fork](https://github.com/bitcoin-core/leveldb-subtree).

<hr /><em>This discussion was created from the release <a href='https://github.com/dogecoin/dogecoin/releases/tag/v1.14.6'>Dogecoin Core 1.14.6</a>.</em>

**Comment by paulVu** [comment]
Thank you all for your hard work

**Reply by ginwigwe200** [reply]
Please I need an app for minning DOGE please you guys should recommend me please

**Comment by patricklodder** [comment]
### Q: What is the difference between a signed and unsigned version?
_Originally posted on [r/dogecoindev](https://www.reddit.com/r/dogecoindev/comments/w41js6/dogecoin_core_1146_released/ih1xe1k/) by [/u/regularqualitysound](https://www.reddit.com/user/regularqualitysound)_

Simple answer: when in doubt, use the signed version.

The signing is basically a mechanism that some OS's use (windows and macOS mainly) for any application that runs on your computer to have some form of anti-malware mechanism (or at least make it harder)

For windows, we currently only have unsigned binaries but I'm working on getting signed ones and hope to add them to the release files.

For macOS, the signed binaries have been notarized by Apple (using my registered developer account) to be unique and untampered. However, there is currently no way to 100% proof from a Dogecoin perspective that that is the same code as what is published on Github. So if you care about such things, then you can download the unsigned version and go through the hassle with Apple's security settings, and then run that. The difference in experience is really: signed = no hassle from Apple because they know I signed it. unsigned = lots of hassle to get it installed but you don't have to trust me at all because you can fully verify on your own that this is the code it says it is.

### Q: Can you confirm that the code signature has changed?
_Originally asked in #3064 by @suschizu_

Correct. Before, the signature was done by @langerhans, and this time it was done by @patricklodder.

**Reply by dave-knight** [reply]
Could you please provide the public key for the private key you used to sign SHA256SUMS.asc?

**Reply by patricklodder** [reply]
Sure! It's maintained under `contrib/gitian-keys`: https://github.com/dogecoin/dogecoin/blob/master/contrib/gitian-keys/patricklodder-key.pgp

**Comment by GiverofMemory** [comment]
Very nice update, thanks for your work!

**Comment by hosensahadat7** [comment]
Yes

**Comment by hosensahadat7** [comment]
Yed

**Comment by hosensahadat7** [comment]
Yes

**Comment by hosensahadat7** [comment]
Yes

**Comment by hosensahadat7** [comment]
Yes

**Comment by hosensahadat7** [comment]
Much wow

**Comment by hosensahadat7** [comment]
Yes

**Comment by Reaper-lk** [comment]
Awesome! could you please tell me what is the build system used for create binaries?

**Reply by patricklodder** [reply]
we use gitian-builder at the moment:

- system: https://github.com/devrandom/gitian-builder
- our descriptors: https://github.com/dogecoin/dogecoin/tree/master/contrib/gitian-descriptors
- the docs: https://github.com/dogecoin/dogecoin/blob/master/doc/gitian-building.md

**Reply by Macomarin** [reply]
Nice one

**Reply by Reaper-lk** [reply]
when try to build. I'm getting this eror. could you please help me? config-bootstrap-fixup file existing 
![image](https://user-images.githubusercontent.com/86299599/216769166-585d1bff-a3ac-4c0b-98a8-8b5b1b87dedb.png)

**Reply by patricklodder** [reply]
I normally run it from one directory underneath and call `bin/make-base-vm` / `bin/gbuild` from there

**Reply by Reaper-lk** [reply]
Thank you! 
![image](https://user-images.githubusercontent.com/86299599/216772386-d0b28265-c96d-4fa0-be55-6a76e79c2788.png)
I failed to resolve this badly. issue with vm builder

**Reply by patricklodder** [reply]
What vm/container host are you targeting? KVM?

Docker is easier to run because there aren't many caveats, and we have a script for it that simplifies it: https://github.com/dogecoin/dogecoin/blob/master/contrib/gitian-build.sh

**Reply by Reaper-lk** [reply]
![image](https://user-images.githubusercontent.com/86299599/216804852-3594ee03-c213-4b2d-82b0-b0175371c6f9.png)
this is after running gitian.sh. what is the next step please?

**Reply by Reaper-lk** [reply]
![image](https://user-images.githubusercontent.com/86299599/216805017-a4b7f26d-4748-4427-9de2-7ade18d3c151.png)

**Reply by Reaper-lk** [reply]
![image](https://user-images.githubusercontent.com/86299599/216806857-b0803d03-883b-402e-a1b4-f96da3484d9a.png)
![image](https://user-images.githubusercontent.com/86299599/216806869-75267f18-a6cc-455f-a95e-cf1e9fbc0377.png)

**Reply by Reaper-lk** [reply]
@patricklodder could you please help me to resolve this?  😥🫠

![image](https://user-images.githubusercontent.com/86299599/216842290-4ebe2b41-496a-44d8-9ed0-f8def6c5167a.png)
![image](https://user-images.githubusercontent.com/86299599/216842336-3854610f-4d11-4052-8833-272544341463.png)

**Reply by patricklodder** [reply]
That's caused by not having `apt-cacher-ng` on your host. See https://github.com/dogecoin/dogecoin/blob/master/doc/gitian-building.md#common-dependencies

**Reply by Reaper-lk** [reply]
Oh thank you! 🔥

**Reply by Reaper-lk** [reply]
@patricklodder 
![image](https://user-images.githubusercontent.com/86299599/217188213-79e3ffb4-4aa6-42ea-82b4-5b2c55d16eea.png)
I'm having this error. could you please help me? Thank you!

**Reply by Reaper-lk** [reply]
when trying with docker

**Reply by patricklodder** [reply]
do you run a firewall like UFW on your host? If so, make sure that docker-to-host traffic on port 3142 is allowed, i.e. `sudo ufw allow from 172.17.0.0/16 to 172.17.0.0/16 port 3142`

**Comment by patricklodder** [comment]
## Security disclosure regarding fixes released with 1.14.6

Today, a [security disclosure](http://halborn.com/blog/post/halborn-discovers-zero-day-impacting-dogecoin-and-280-networks) has been made by [Halborn](https://twitter.com/HalbornSecurity), announcing that critical vulnerabilities were found in Dogecoin Core 1.14.5 software and older. The vulnerabilities were subsequently found to also exist in another 280 instances of bitcoin-derived blockchain software.

Dogecoin maintainers were privately informed by Halborn about the vulnerabilities and resolved these in code that was released with version 1.14.6. All reported critical issues were confirmed to be resolved with the release of [Dogecoin Core 1.14.6](https://github.com/dogecoin/dogecoin/releases/tag/v1.14.6) by Halborn's engineers. The inclusion of the bugfixes were done in "stealth mode", to buy time for the network to upgrade organically and to allow security researchers time to address this with other software engineers and teams. Maintainers were already working at hardening the network code at the time, and we incorporated these fixes into that effort.

Because the vulnerabilities were found in the code that implements the peer-to-peer protocol, individual nodes with low memory (2GB or less) were vulnerable to running out of memory. Today, with over 50% of the network having upgraded, the risk to the network as a whole is believed to be mitigated, but individual nodes that have not yet upgraded can still be vulnerable and all node operators are recommended to upgrade at their earliest convenience. 

If you run an older version of the Dogecoin Core software than 1.14.6, please read the upgrade guidance below to help you updating your software.

### Upgrade guidance

If you're currently running Dogecoin Core 1.8.x, 1.10.0, 1.14.0 or 1.14.5, an upgrade to 1.14.6 is straight-forward. Just install the updated binaries for your platform from the [release page](https://github.com/dogecoin/dogecoin/releases/tag/v1.14.6) and restart the software. You can identify the version you're running in the bottom lower corner of the Dogecoin Qt home screen, or with `dogecoind -version`.

If you're running either of versions 1.14.2, 1.14.3 or 1.14.4, you use the wallet functionality, AND you have in the past used 1.14.0, unexpected side-effects from a policy bug fix in version 1.14.2 may have caused your wallet to currently have stuck transactions. An upgrade to 1.14.6 could make these transactions "unstuck" and send them out, and it is recommended to either purge all pending wallet transactions, or manually check for stuck transactions:

#### Purging all pending wallet transactions automatically

To automatically purge all pending transactions, restart `dogecoind` or `dogecoin-qt` once, with the `-zapwallettxes=1` option, eg:

```console
$ dogecoind -zapwallettxes=1

> Dogecoin-Qt.exe -zapwallettxes=1
```

#### Manually searching for and removing pending wallet transactions

To manually check and remove pending transactions, use the `listtransactions` RPC call, either using dogecoin-cli, or from the Debug console. Look for transactions with 0 confirmations, or filter them with `jq`, like so:

```console
$ dogecoin-cli listtransactions "*" 100000 0 | jq -r '.[] | select(.confirmations == 0)'
{
  "account": "",
  "address": "D9gEMMtCNT5GHvFvSQMjfSnbZFRRUjjDPX",
  "category": "send",
  "amount": -420.69,
  "vout": 0,
  "confirmations": 0, <---- this means the transaction has never been mined
  "trusted": true,
  "txid": "47a84a30159ed15d36e91ef6ae161fe5254aa9c2c906dc614d5aba03659e65ed",
  "walletconflicts": [],
  "time": 1678343421,
  "timereceived": 1678343421,
  "bip125-replaceable": "no",
  "abandoned": false <---- this means the transaction is stuck and has not been purged from the wallet
} 
```

These transactions can then be removed with `abandontransaction` individually, if desired:

```console
$ dogecoin-cli abandontransaction 47a84a30159ed15d36e91ef6ae161fe5254aa9c2c906dc614d5aba03659e65ed
```

#### Post-upgrade checks

After abandoning transactions, it should be safe to update the software to version 1.14.6. If you want to double check after you upgraded the software, you can mimic the 1.14.2 dust relay rules by setting the hard dust limit when starting dogecoind or dogecoin-qt with `-harddustlimit=1`. This sets the minimum amount of DOGE in a transaction output to 1.0 and rejects any transactions that attempt to transfer lower amounts. Alternatively, you can run the software once while not connected to the internet. 1.14.6 ships with a new `liststucktransactions` RPC method that can be used to check if there are still stuck transactions in the wallet:

```console
$ src/dogecoin-cli liststucktransactions
[]
```

**Reply by Reaper-lk** [reply]
@patricklodder Could you please provide more information about the commits related to Halborn's issue?

**Reply by patricklodder** [reply]
I only saw this question now - my apologies. Per ethical disclosure standards, I will not provide further information until the CVE is updated with the actual details, to protect other blockchain software from being exploited by details provided by us on this codebase.

Once I get a notification from Halborn that they will move to the next stage, I will be able to provide further information. Please contact Halborn through their website if you're asking this because you are maintaining a Bitcoin Core software fork and you worry that your software is exposed. They have developed tools to quickly determine vulnerability.

**Comment by VoytekMich** [comment]
$ dogecoin-cli getnetworkinfo
`{
  "version": 1140500,
  "subversion": "/Shibetoshi:1.14.5/",
  "protocolversion": 70015,
...`

file [dogecoin-1.14.6-x86_64-linux-gnu.tar.gz](https://github.com/dogecoin/dogecoin/releases/download/v1.14.6/dogecoin-1.14.6-x86_64-linux-gnu.tar.gz)

Something wrong with file or bug in core?
Version should be 1140600 , right?

**Reply by patricklodder** [reply]
I double-checked the GitHub hosted binary and the files are fine:

```bash
curl -L https://raw.githubusercontent.com/dogecoin/dogecoin/master/contrib/gitian-keys/patricklodder-key.pgp | gpg --import
curl -L -O https://github.com/dogecoin/dogecoin/releases/download/v1.14.6/dogecoin-1.14.6-x86_64-linux-gnu.tar.gz
curl -L -O https://github.com/dogecoin/dogecoin/releases/download/v1.14.6/SHA256SUMS.asc
gpg --verify SHA256SUMS.asc
grep dogecoin-1.14.6-x86_64-linux-gnu.tar.gz SHA256SUMS.asc | shasum -c
```

Then checked the binary from the tarball:

```console
$ tar zxfv dogecoin-1.14.6-x86_64-linux-gnu.tar.gz
$ dogecoin-1.14.6/bin/dogecoind -regtest -daemon
Dogecoin server starting
$ dogecoin-1.14.6/bin/dogecoin-cli -regtest getnetworkinfo
{
  "version": 1140600
[..]
$ dogecoin-1.14.6/bin/dogecoin-cli -regtest stop
Dogecoin server stopping
```

I assume that you're calling an old binary when you start the daemon. If you're just starting with `dogecoind -daemon`, you may have the 1.14.5 binary somewhere in your path - find with `which dogecoind`.
